mypy/options.py

```diff
@@ -208,27 +210,54 @@ def __ne__(self, other: object) -> bool:
     def __repr__(self) -> str:
         return 'Options({})'.format(pprint.pformat(self.snapshot()))
 
+    def apply_changes(self, changes: Dict[str, object]) -> 'Options':
+        new_options = Options()
+        new_options.__dict__.update(self.__dict__)
+        new_options.__dict__.update(changes)
+        return new_options
+
     def build_per_module_cache(self) -> None:
         self.per_module_cache = {}
-        # Since configs inherit from glob configs above them in the hierarchy,
+
+        # Config precedence is as follows:
+        #  1. Concrete section names: foo.bar.baz
+        #  2. "Unstructured" glob patterns: foo.*.baz, in the order they appear in the file
+        #  3. "Well-structured" wildcard patterns: foo.bar.*, in specificity order.
+
+        # Since structured configs inherit from glob configs above them in the hierarchy,
         # we need to process per-module configs in a careful order.
-        # We have to process foo.* before foo.bar.* before foo.bar.
-        # To do this, process all glob configs before non-glob configs and
+        # We have to process foo.* before foo.bar.* before foo.bar,
+        # and we need to apply *.bar to foo.bar but not to foo.bar.*.
+        # To do this, process all well-structured glob configs before non-glob configs and
         # exploit the fact that foo.* sorts earlier ASCIIbetically (unicodebetically?)
         # than foo.bar.*.
-        keys = (sorted(k for k in self.per_module_options.keys() if k.endswith('.*')) +
-                [k for k in self.per_module_options.keys() if not k.endswith('.*')])
-        for key in keys:
+        # Unstructured glob configs are stored and are all checked for each module.
+        unstructured_glob_keys = [k for k in self.per_module_options.keys()
+                                  if '*' in k[:-1]]
+        structured_keys = [k for k in self.per_module_options.keys()
+                           if '*' not in k[:-1]]
+        wildcards = sorted(k for k in structured_keys if k.endswith('.*'))
+        concrete = [k for k in structured_keys if not k.endswith('.*')]
+
+        for glob in unstructured_glob_keys:
+            self.glob_options.append((glob, re.compile(fnmatch.translate(glob))))
```

Comment:
IIUC this means that module `foo.bar` does not match the pattern `foo.*.bar`. And that's surprising since `foo.bar` does matchs `foo.bar.*`.